google/cloud/spanner/mutations.h

```diff
@@ -0,0 +1,234 @@
+ * using a data structure instead of a SQL statement.
+ *
+ * This class serves as a wrapper for all mutations types. Use the builders,
+ * such as `InsertMutationBuilder` or `UpdateMutationBuilder` to create objects
+ * of this class.
+ *
+ * @see https://cloud.google.com/spanner/docs/modify-mutation-api
+ *   for more information about the Cloud Spanner mutation API.
+ */
+class Mutation {
+ public:
+  /**
+   * Creates an empty mutation.
+   *
+   * @note Empty mutations are not usable with the Cloud Spanner mutation API.
+   */
+  Mutation() = default;
+
+  Mutation(Mutation&&) noexcept = default;
+  Mutation& operator=(Mutation&&) noexcept = default;
+  Mutation(Mutation const&) = default;
+  Mutation& operator=(Mutation const&) = default;
+
+  friend bool operator==(Mutation const& lhs, Mutation const& rhs);
+  friend bool operator!=(Mutation const& lhs, Mutation const& rhs) {
+    return !(lhs == rhs);
+  }
+
+  /// Convert the mutation to the underlying proto.
+  google::spanner::v1::Mutation as_proto() && { return std::move(m_); }
+
+  /**
+   * Allows Google Test to print internal debugging information when test
+   * assertions fail.
+   *
+   * @warning This is intended for debugging and human consumption only, not
+   *   machine consumption as the output format may change without notice.
+   */
+  friend void PrintTo(Mutation const& m, std::ostream* os);
+
+ private:
+  friend class InsertMutationBuilder;
+  friend class UpdateMutationBuilder;
+  explicit Mutation(google::spanner::v1::Mutation&& m) : m_(std::move(m)) {}
+
+  google::spanner::v1::Mutation m_;
+};
+
+// This namespace contains implementation details. It is not part of the public
+// API, and subject to change without notice.
+namespace internal {
+inline void PopulateListValue(google::protobuf::ListValue&) {}
+
+/**
+ * Initialize a `google::protobuf::ListValue` from a variadic list of C++ types.
+ *
+ * This function applies the conversions defined by #Value
+ */
+template <typename T, typename... Ts>
+void PopulateListValue(google::protobuf::ListValue& lv, T&& head,
+                       Ts&&... tail) {
+  google::spanner::v1::Type type;
+  google::protobuf::Value value;
+  std::tie(type, value) = internal::ToProto(Value(std::forward<T>(head)));
+  *lv.add_values() = std::move(value);
+  PopulateListValue(lv, std::forward<Ts>(tail)...);
+}
+
+template <typename Op>
+class WriteMutationBuilder {
+ public:
+  WriteMutationBuilder() { Op::mutable_field(m_); }
+
+  explicit WriteMutationBuilder(std::vector<std::string> column_names) {
+    auto& field = Op::mutable_field(m_);
+    field.mutable_columns()->Reserve(column_names.size());
+    for (auto& name : column_names) {
+      *field.add_columns() = std::move(name);
+    }
+  }
+
+  google::spanner::v1::Mutation Build() const& { return m_; }
+  google::spanner::v1::Mutation Build() && { return std::move(m_); }
+
+  template <typename... Ts>
+  void AddRow(Ts&&... values) {
```

Comment:
I wonder if things where we "add rows" should accept a `Row<Ts...>` maybe instead of, or in addition to, `Value` objects? I guess it depends on the use case, what type callers may be more likely to have. For example, I could imagine that it may be nice to write this like:

```cc
builder.AddRow(MakeRow("hello", 42, "world", true));
```
As opposed to
```cc
builder.AddRow(Value("hello"), Value(42), Value("world"), Value(true));
```

However, if we go toward accepting `Row`, we may want to wait for a PR I just sent #214 because then it'll be easier to extract the `Value` objects back out of the Row.

Reply:
I think we should go with an overload that takes `Row<Ts...>`. Not all applications will have a `Row` handy, the data they want to put into a mutation may come from anywhere.